fix: App crashes on simultaneous TopAppBar and system navigation back button press

easyhooon opened the pull request:
## Issue
- close https://github.com/DroidKaigi/conference-app-2025/issues/264

## Overview (Required)
- Prevents "java.lang.IllegalArgumentException: NavDisplay backstack cannot be empty"
  when rapidly tapping back buttons in navigation3 by adding type-safe removal functions.
- ~~Replace type-specific back handling with MainScreenTab-based approach.
This fixes UI-level back handling on SearchScreen and other non-main screens
while still preventing backstack empty crashes on rapid tapping.~~

## Links
- https://blog.csdn.net/qq_50675668/article/details/149772402
- https://github.com/android/androidify/pull/13
- https://issuetracker.google.com/issues/419017866

## Movie (Optional)
Before | After
:--: | :--:
<video src="https://github.com/user-attachments/assets/307bbcb4-ead9-4252-b059-d8a95ff3a763" width="300" > | <video src="https://github.com/user-attachments/assets/27d8f733-5b4b-462f-988d-310529139bf6" width="300" >

easyhooon committed:
fix: prevent backstack empty error on rapid tapping

Prevents "java.lang.IllegalArgumentException: NavDisplay backstack cannot be empty"
  when rapidly tapping back buttons in navigation3

easyhooon commented on app-shared/src/androidJvmMain/kotlin/io/github/droidkaigi/confsched/KaigiAppUi.androidJvm.kt:
```diff
@@ -71,7 +72,7 @@ actual fun KaigiAppUi() {
             sceneStrategy = sceneStrategy(),
             entryProvider = entryProvider {
                 sessionEntries(
-                    onBackClick = { backStack.removeLastOrNull() },
+                    onBackClick = { backStack.safeRemoveLastIfType<TimetableItemDetailNavKey>() },
```
Ah, I missed another sideEffect ;). In this case, I think we can modify it to execute based on whether the remaining item in the backStack is a NavKey that belongs to MainBottomTab.

easyhooon replied:
https://github.com/user-attachments/assets/0972e22f-b560-40d8-8a5e-114731982825

easyhooon committed:
fix: improve back navigation to preserve main tab keys

Replace type-specific back handling with MainScreenTab-based approach.
This fixes UI-level back handling on SearchScreen and other non-main screens
while still preventing backstack empty crashes on rapid tapping.

easyhooon committed:
chore: apply spotless

kitakkun reviewed (commented):
Thank you for updating the code!
Please take a look at my comments when you have a chance.

easyhooon committed:
refactor: simplify back navigation logic per code review

Simplified safeRemoveLastPreservingRoot to safeRemoveLastOrNull by only checking size > 1.
Removed unused mainTabNavKeys and navKey extension as root key preservation is no longer needed.
Updated comment to clarify the issue occurs when system and UI back buttons are pressed simultaneously.

easyhooon commented on app-shared/src/androidJvmMain/kotlin/io/github/droidkaigi/confsched/navigation/extension/BackStackExtensions.kt:
```diff
@@ -0,0 +1,43 @@
+    } else {
+        null
+    }
+}
```
When I first worked on this issue, checking only by size didn't resolve the app crash problem, so I added logic to check a separate key. But now when I check again, it works fine without any issues...! Code review feedback has been implemented!

easyhooon replied:
https://github.com/android/androidify/pull/13
https://issuetracker.google.com/issues/419017866

I was able to confirm that reports like these issue have already been registered. While there was a comment saying it was resolved in the latest version of navigation3 and that no additional code needed to be added, it seems that even after upgrading to the current latest version (alpha07), the issue still persists when simultaneously pressing the system back button and the topAppBar back button...

kitakkun replied:
Navigation3 is still in alpha, so the solution doesn’t need to be perfect. This already lowers the chance of crashes 🎉
I’m happy to merge this! Let me know if you’d like to refine it a bit more, otherwise I can go ahead and merge.

kitakkun commented on app-shared/src/androidJvmMain/kotlin/io/github/droidkaigi/confsched/KaigiAppUi.androidJvm.kt:
Apologies, I missed this earlier — it looks like we might also need to apply a similar fix to aboutEntries’s onBackClick 🙏🏻 The crash was reproducible on the AboutScreen.

easyhooon committed:
fix: apply safeRemoveLastOrNull to aboutEntries onBackClick

kitakkun reviewed (approved):
Thanks for making the adjustments. On to the merge 🚀

easyhooon commented:
Maybe it can be solved by upgrading navigation3 version 1.0.0-rc01!

https://developer.android.com/jetpack/androidx/releases/navigation3#1.0.0-rc01
